### ISSUE TITLE
Http/2 | ws over tls transport support and proxy through caddy/nginx

### ISSUE THREAD

**Issue by bash99** [blinksocks/blinksocks#103]
- [ ] feature request
Can we support Http/2 + path like v2ray, so a normal website can co-exists with blinksocks.

As request to Yahoo or Bing is http/2 now, I think it will much harder for GFW to do traffic analysis.

Have try ws over caddy but failed, (direct ws is ok)
Caddyfile:
```
    proxy /blinksocks 127.0.0.1:48964 {
        websocket
    }
```

Blink Server:
```
{
  "service": "ws://0.0.0.0:48964/blinksocks",
  "key": "my-key",
  "presets": [
    {
      "name": "ss-base"
    },
    {
      "name": "obfs-random-padding"
    },
    {
      "name": "ss-stream-cipher",
      "params": {
        "method": "aes-128-ctr"
      }
    }
  ]
}
```

Blink Client
```

{
  "service": "socks5://127.0.0.1:43779",
  "server": {
    "service": "ws://my-server:443/blinksocks",
    "key": "my-key",
    "presets": [
      {
        "name": "ss-base"
      },
      {
        "name": "obfs-random-padding"
      },
      {
        "name": "ss-stream-cipher",
        "params": {
          "method": "aes-128-ctr"
        }
      }
    ]
  }
}
```

**Comment by micooz** [comment]
1. HTTP/2 transport is working in progress in branch [feature-http2-transport](https://github.com/blinksocks/blinksocks/tree/feature-http2-transport). But I don't think it's much harder for GFW to do traffic analysis, transports are just clumsy modification to TCP, traffic analysis aims at a different dimension.
2. Any details of error message? Client/Server/Caddy logs?

**Comment by shengshampoo** [comment]
Read someone's simplified chinese article below:
[使用Caddy反向代理Project V](https://ellinia.me/Use_caddy_to_reverse_proxy_project_v/)

Referring to his config files about caddyfile, 
I could connects  successfully to blinsocks client with blinksocks server through ws and proxy through caddy.

PS: caddyfile  **"proxy /blinksocks"** url is relevant to blinksocks.server.json **"ws://0.0.0.0:port/blinksocks"**.

* Caddyfile
```
custom_domain.com:port {
root /path/
gzip
forwardproxy {
        basicauth       PROXY_USER      PROXY_PASS
}
proxy /blinksocks localhost:ws_the_same_port {
       websocket
       header_upstream Connection {>Connection}
       header_upstream Upgrade {>Upgrade}
       header_upstream Host {Host}
    }
}
```

blinksocks.client.json
```
"service": "socks5://127.0.0.1:1081",
  "server": {
    "service": "ws://vps_ip:port/blinksocks",
    "key": "my-key",
   ......................
   },
  ......................
  .....................
```

blinksocks.server.json
```
{
  "service": "ws://0.0.0.0:port/blinksocks",
  "key": "my-key",
  ..................
  ..................
},

**Comment by bash99** [comment]
@shengshampoo 
Your blinksocks.client.json still use ws_port to blink server, so caddy didn't play a role in it.
    "service": "ws://vps_ip:ws_port/blinksocks",

if you connect to 443 and hope caddy pass through it, will it works?
    "service": "ws://vps_ip:443/blinksocks",

**Comment by shengshampoo** [comment]
Look back another documents about nginx/caddy reverse proxy and [v2ray official tutorial][1],
I konw some mistakes in my config files and correct some points as above. 

* Caddyfile
 ~custom_domain:port~ => https://custom_domain:443

* blinksocks.client.json
 ~ws://vps_ip:ws_port/blinksocks~ => ws://custom_domain:443/blinksocks

Maybe it is lucky for me that new newbie could let it connect successfully, 
or is it still not completely correct answer?

[1]: https://toutyrater.github.io/advanced/wss_and_web.html

```

**Comment by bash99** [comment]
@micooz Hi, I've try v3.2.2 and found blinksocks send data as (capture by strace)
```
[pid 24739] read(29, "GET /blink.path HTTP/1.1\r\nSec-WebSocket-Version: 13\r\nSec-WebSocket-Key: 4UF59C44RwXNHEpbwK4BVw==\r\nConnection: Upgrade\r\nUpgrade: websocket\r\nHost: myhost.com:443\r\n\r\n", 1024) = 166
```
, as it's not tls, caddy close it asap.
```
[pid 24739] write(29, "\25\3\1\0\2\2\n", 7) = 7
[pid 24739] epoll_ctl(4, EPOLL_CTL_DEL, 29, c4209adaf4) = 0
[pid 24739] close(29)                   = 0
```

I've compare with v2ray/ws/tls setting (https://github.com/KiriKira/vTemplate/tree/master/websocket%2BCaddy%2BTLS(use%20path) )
which v2ray send a sni with hostname:
```
[pid 25244] read(26, "\26\3\1\1?\1\0\1;\3\3\214\344\203V\317\34\260\324Op\10\33~\322\31i\237\234N\273\242\302\226\204N%\235foP\247\277\20\330\257%\227\20\256\rm\302\230\275t\227\231k\202\0 \300/\3000\300+\300,\314\250\314\251\300\23\300\t\300\24\300\n\0\234\0\235\0/\0005\300\22\0\n\1\0\0\3423t\0\0\0\0\0\23\0\21\0\0\016myhost.com\0\5\0\5\1\0\0\0\0\0\n\0\n\0\10\0\35\0\27\0\30\0\31\0\v\0\2\1\0\0#\0x\31\23\233\267\254\230\266\23\377\205;\226\310R\316\263\323\325\23\366_\360+\257\247\255F\346\365\252\233\333\253\325\34yJ\244c-s\243\227F\312\354\351[\235\312\31P\370\352\214>\362\207\362{\200s\360\241\16!\312\27@\215\36\323i\255%&\24\6\323\371\236\220\261\336d\226\f\322I\316\310\272\211\273\346\205\270\267I^\364\364iM\362\377<\363\371\243\214\252\247\256\2136\373\37 E\0\r\0\22\0\20\4\1\4\3\5\1\5\3\6\1\6\3\2\1\2\3\377\1\0\1\0\0\20\0\v\0\t\10http/1.1\0\22\0\0", 1024) = 324
```

So I think blinksocks should do a tls handshake first ?

**Comment by micooz** [comment]
Hi, @bash99 I just published a new version([v3.3.0](https://github.com/blinksocks/blinksocks/releases/tag/v3.3.0)) to npm registry which add [WebSocket/TLS](https://github.com/blinksocks/blinksocks/tree/master/docs/config#blinksocks-over-websockettls) support.

For your scenario, keep `ws://` in **server.json**, change **client.json** to use `wss://` protocol:

```
[client]  <-> [caddy] <-> [server] <-> [target]
  wss://        :443        ws://
```

Please test on this version, and let me know if you succeeded.

**Comment by shengshampoo** [comment]
For new version(v3.3.0) which add **WebSocket/TLS support**,
There are some brief records with my test.

```
[client]  <-> [server] <-> [target]
  wss://        wss://
```

as following configs between could **let it connect successfully.**

* blinksocks.wss.client.json

  1. if mux enable, blinksocks client program will immediately abort.
  2. It is not required to key in
     
    ```
   "tls_cert": "cert.pem",
   "tls_cert_self_signed": false,
    ```
     Maybe the reason is that use caddy automatic generated let's encrypt cert.

* blinksocks.wss.client.json

```
{
  "service": "socks5://127.0.0.1:1081",
  "server": {
    "service": "wss://custom_domain:wss_port",
    "key": "",
     .................
     .................
    "mux": false, // if mux enable, will terminate client program abnormally.
    "mux_concurrency": 10
  },
     .........................
}
```
	
* blinksocks.wss.server.json

```
{
  "service": "wss://0.0.0.0:wss_port",
  "tls_key": "/path/.caddy/custom_domain_let's encrypt self signed.key",
  "tls_cert": "/path/.caddy/custom_domain_let's encrypt self signed.crt",
  "key": "",
   ........................
   ........................
  "mux": false,
  ......................
}
```

**Comment by shengshampoo** [comment]
However, 
ever if on new version 3.3.0, 
still couldn't find the correct configs for blinksocks and caddy.

* blinksocks.client.json with **wss**

```
"service": "socks5://127.0.0.1:1081",
  "server": {
    "service": "wss://custom_domain:443/blinksocks",
    "key": "my-key",
   ......................
   },
  ......................
  .....................

```

only when port is **443** , blinksocks client execute failed and show error as follwing: 

```
Error: service.port is invalid
    at Function._testServer (/usr/lib/node_modules/blinksocks/lib/core/config.js:355:13)
    at Function.testOnClient (/usr/lib/node_modules/blinksocks/lib/core/config.js:288:12)
    at Function.test (/usr/lib/node_modules/blinksocks/lib/core/config.js:228:14)
    at bootstrap (/usr/lib/node_modules/blinksocks/bin/bootstrap.js:31:12)
    at main (/usr/lib/node_modules/blinksocks/bin/cli.js:100:12)
    at Object.<anonymous> (/usr/lib/node_modules/blinksocks/bin/cli.js:113:1)
    at Module._compile (internal/modules/cjs/loader.js:702:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:713:10)
    at Module.load (internal/modules/cjs/loader.js:612:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:551:12)

```

**Comment by micooz** [comment]
@shengshampoo **mux** will not work of course, because caddy cannot recognize mux protocol, it's totally different from ws(http)/wss(https).

@bash99 thx for your feedback, I'll fix it soon. BTW, can you share all your configs here again so that I can make an example list in the document.

**Comment by bash99** [comment]
@micooz full config is below
server.json
```
{
  "service": "ws://0.0.0.0:40443/xxx.blk",
  "key": "mykey",
  "presets": [
    {
      "name": "ss-base"
    },
    {
      "name": "obfs-random-padding"
    },
    {
      "name": "ss-stream-cipher",
      "params": {
        "method": "aes-128-ctr"
      }
    }
  ]
}
```
client.json
```
{
  "service": "socks5://127.0.0.1:1080",
  "server": {
    "service": "wss://myserver.com:443/xxx.blk",
    "key": "mykey",
    "presets": [
      {
        "name": "ss-base"
      },
      {
        "name": "obfs-random-padding"
      },
      {
        "name": "ss-stream-cipher",
        "params": {
          "method": "aes-128-ctr"
        }
      }
    ]
  }
}
```

caddy:
```
myserver.com {
    root /var/www/html
    gzip
    proxy /xxx.blk 127.0.0.1:40443 {
        websocket
        header_upstream -Origin
    }
}
```

**Comment by micooz** [comment]
I just published [v3.3.1](https://github.com/blinksocks/blinksocks/releases/tag/v3.3.1) and made a successful test with caddy, and put some examples [here](https://github.com/blinksocks/blinksocks/tree/v3.3.1/docs/examples).

**Comment by shengshampoo** [comment]
Nice,  also could use  this version 3.3.1 to support websocket + TLS + caddy successful.
As following simplified chinese  blog:
[v2ray抗封锁最优配置](https://www.feiqy.com/v2ray-best-config/)
For ISP, such as China union, this  network traffic look as normal app browsing on it's network statistics.

**Comment by micooz** [comment]
Hi all, I just published another version [3.3.2] which support [HTTP2] and [HTTP2+Caddy]. Multiplexing should also work when using wss or http2. 😃

[3.3.2]: https://github.com/blinksocks/blinksocks/releases/tag/v3.3.2
[HTTP2]: https://github.com/blinksocks/blinksocks/tree/master/docs/examples/http2
[HTTP2+Caddy]: https://github.com/blinksocks/blinksocks/tree/master/docs/examples/http2-caddy

**Comment by shengshampoo** [comment]
I think HTTP/2 is better than websocket+tls because of websocket is old HTTP 1.1 protocol, 
and few CDN ( only cloudFlare ?) support websocket to hide website's IP Address and speed up Internet connection.
